pkg/utils/gwlog/metadata.go

```diff
@@ -0,0 +1,86 @@
+package gwlog
+
+import (
+	"context"
+	"github.com/google/uuid"
+)
+
+type key string
+
+const metadata key = "metadata"
+
+const traceID string = "trace_id"
+
+type metadataValue struct {
+	m map[string]string
+}
+
+func (mv *metadataValue) get(key string) string {
+	return mv.m[key]
+}
+
+func (mv *metadataValue) set(key, val string) {
+	mv.m[key] = val
+}
+
+func newMetadata() *metadataValue {
+	return &metadataValue{
+		m: make(map[string]string),
+	}
+}
+
+func NewTrace(ctx context.Context) context.Context {
+	currID := uuid.New()
+
+	newCtx := context.WithValue(ctx, metadata, newMetadata())
+	AddMetadata(newCtx, traceID, currID.String())
+
+	return newCtx
+}
+
+func AddMetadata(ctx context.Context, key, value string) {
+	if ctx.Value(metadata) != nil {
+		ctx.Value(metadata).(*metadataValue).set(key, value)
+	}
+}
+
+func GetMetadata(ctx context.Context) []interface{} {
+	var fields []interface{}
+
+	if ctx.Value(metadata) != nil {
+		for k, v := range ctx.Value(metadata).(*metadataValue).m {
+			if k == traceID {
+				// skip since there's a separate method to grab the trace id
+				continue
+			}
+			fields = append(fields, k)
+			fields = append(fields, v)
+		}
+	}
+	return fields
+}
+
+func GetTrace(ctx context.Context) string {
+	if ctx.Value(metadata) != nil {
+		m := ctx.Value(metadata).(*metadataValue).m
+		if m == nil {
+			return ""
+		}
+		return ctx.Value(metadata).(*metadataValue).m[traceID]
+	}
+	return ""
+}
+
+func StartReconcileTrace(ctx context.Context, log Logger, k8sresourcetype, name string) context.Context {
+	ctx = NewTrace(ctx)
+	AddMetadata(ctx, "type", k8sresourcetype)
+	AddMetadata(ctx, "name", name)
+
+	log.Infow(ctx, ReconcileStart, GetMetadata(ctx)...)
```

Comment:
Rather than a special method to exclude the trace id from metadata, what about trying to cast to a `TracedLogger`, then either calling a method that logs all metadata, including the trace id, or logging directly against the `InnerLogger`. If it's not a `TracedLogger`, then we can just log all metadata.

Reply:
TracedLogger and Innerlogger are not interchangable. They do not implement the same interface since I added a context parameter to the methods of the original logger. That's unfortunately why I had to address every usage of the logger.

I think that would be quite cool but I may be missing how this might work. Maybe you can give an example.

Reply:
So here I was thinking of something like
```
traceLogger, ok := log.(TraceLogger)
if ok {
  // here, getMetadata would include everything including trace ID
  traceLogger.InnerLogger.Infow(ReconcileStart, getMetadata(ctx))
}
```
but this was because I missed we were constructing a list rather than a map without the trace id. So I don't think my idea really saves anything other than excluding the trace id from the metadata list.